### PR TITLE
chore(compiler): avoid storing types on AST nodes

### PR DESCRIPTION
**Pull request by Chriscbr**
Decouple type information from AST nodes by storing them directly inside the `Types` struct, linked through IDs that are unique to every expression. This cleans up the usages of `expression.evaluated_type.borrow().unwrap()` that have been required each time the type of an expression was needed.

No tests or docs updated since there weren't any functional changes.

## Checklist

- [x] Title matches [Winglang's style guide](https://docs.winglang.io/contributors/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -1086,6 +1089,30 @@ impl Types {
...
+	}
+
+	/// Obtain the type of a given expression node. Returns None if the expression has not been type checked yet. If
+	/// this is called after type checking, it should always return Some.
+	pub fn get_expr_type(&self, expr: &Expr) -> Option<TypeRef> {
+		let expr_idx = expr.id;
+		if self.type_for_expr.len() <= expr_idx {
+			return None;
+		}
+		self.type_for_expr[expr_idx]
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:1109]: nit: this might work and be cleaner
```suggestion
		self.type_for_expr.get(expr_idx).map(|t| *t);
```

**Review by yoav-steinberg — APPROVED** [review]
Love it!

File: libs/wingc/src/ast.rs [hunk 2]
```diff
@@ -7,7 +7,11 @@ use indexmap::{Equivalent, IndexMap, IndexSet};
 
 use crate::diagnostic::WingSpan;
 use crate::type_check::symbol_env::SymbolEnv;
-use crate::type_check::TypeRef;
+
+// assumption: the compiler's main phases (parsing, type checking, jsifying) are executed in a single-threaded context
+thread_local! {
+	static EXPR_COUNTER: RefCell<usize> = RefCell::new(0);
+}
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingc/src/ast.rs:14]: Sure, let me give it a try

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:1109]: nice catch :) That snippet produced an Option<Option<TypeRef>>, but using `.map().flatten()` aka `.and_then()` fixed it

File: libs/wingc/src/ast.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingc/src/ast.rs]: Done

**Review by MarkMcCulloh — APPROVED** [review]

**Review by eladb — COMMENTED** [review]
🙏

**Pull request merged**